lib/rouster.rb

```diff
@@ -292,60 +293,87 @@ def inspect
   #
   # runs a command inside the Vagrant VM
   #
-  # returns output (STDOUT and STDERR) from command run, sets @exitcode
-  # currently determines exitcode by tacking a 'echo $?' onto the command being run, which is then parsed out before returning
+  # returns output (STDOUT and STDERR) from command run, pushes to self.ssh_exitcode
   #
   # parameters
   # * <command> - the command to run (sudo will be prepended if specified in object instantiation)
   # * [expected_exitcode] - allows for non-0 exit codes to be returned without requiring exception handling
-  def run(command, expected_exitcode=[0])
+  def run( command, expected_exitcode = 0, sudo = self.uses_sudo? )
+
+    cmd = {
+      :command           => command,
+      :sudo              => sudo,
+      :stdout            => String.new,
+      :stderr            => String.new,
+      :expected_exitcode => Array( expected_exitcode ),
+      :exitcode          => nil,
+    }
+
+    cmd[:final_command] = cmd[:sudo] ? sprintf( 'sudo bash -c "%s"', command ) : command
 
     if @ssh.nil?
       self.connect_ssh_tunnel
     end
 
-    output = nil
-    expected_exitcode = [expected_exitcode] unless expected_exitcode.class.eql?(Array) # yuck, but 2.0 no longer coerces strings into single element arrays
-
-    cmd = sprintf('%s%s; echo ec[$?]', self.uses_sudo? ? 'sudo ' : '', command)
-    @logger.info(sprintf('vm running: [%s]', cmd)) # TODO decide whether this should be changed in light of passthroughs.. 'remotely'?
+    @logger.info( sprintf( 'vm running: [%s]', cmd[:final_command] ) ) # TODO decide whether this should be changed in light of passthroughs.. 'remotely'?
 
     0.upto(@retries) do |try|
       begin
         if self.is_passthrough? and self.passthrough[:type].eql?(:local)
-          output = `#{cmd}`
+          cmd[:stdout]   = `#{cmd[:final_command]}`
+          cmd[:exitcode] = $?
         else
-          output = @ssh.exec!(cmd)
+          cmd = remote_exec( cmd )
         end
-
         break
       rescue => e
-        @logger.error(sprintf('failed to run [%s] with [%s], attempt[%s/%s]', cmd, e, try, retries)) if self.retries > 0
+        @logger.error(sprintf('failed to run [%s] with [%s], attempt[%s/%s]', cmd[:final_command], e, try, retries)) if self.retries > 0
         sleep 10 # TODO need to expose this as a variable
       end
-
     end
 
-    if output.nil?
-      output    = "error gathering output, last logged output[#{self.get_output()}]"
-      @exitcode = 256
-    elsif output.match(/ec\[(\d+)\]/)
-      @exitcode = $1.to_i
-      output.gsub!(/ec\[(\d+)\]\n/, '')
-    else
-      @exitcode = 1
+    if cmd[:stdout].nil?
+      cmd[:stdout]   = "error gathering output, last logged output[#{self.get_ssh_stdout()}]"
```

Comment:
not sure it makes sense to include `self.get_ssh_stdout()` if the block here will only trigger if it is `nil`.. maybe change this to `.get_ssh_stderr()`? although setting that to be `:stdout` might be more confusing.